### ISSUE TITLE
CNTK AdaDelta Python vs C# performance issues

### ISSUE THREAD

**Issue by DomnenkoB** [Microsoft/CNTK#3367]
I am using [AdaDelta ](https://cntk.ai/pythondocs/cntk.learners.html#cntk.learners.adadelta)as an optimizer for a neural network (XOR-problem, 1 hidden layer, TanH as an activation function, output layer activation function is sigmoid)

```
INPUT_DIM = 2
OUTPUT_DIM = 1
input_ = C.input_variable(INPUT_DIM, 'float32')
output_ = C.input_variable(OUTPUT_DIM, 'float32')

with C.layers.default_options(activation=C.tanh, init=C.glorot_uniform()):
    model = C.layers.Sequential([
        C.layers.Dense(4),
        C.layers.Dense(OUTPUT_DIM, activation=C.sigmoid)
    ])
    model = model(input_)
```

C# neural network configuration is similar. Python and C# have different parameters for AdaDelta.

In Python it's simple as that, most of the parameters have default values
`learner = C.adadelta(model.parameters)`

While in C# I set them explicitly, according to python documentation and it looks like this

double learningRate = 1;
var schedule = new TrainingParameterScheduleDouble(learningRate);

```
var options = new AdditionalLearningOptions()
{
    l1RegularizationWeight = 0,
    l2RegularizationWeight = 0,
    gradientClippingThresholdPerSample = double.PositiveInfinity,
    gradientClippingWithTruncation = true
};

var parameterVector = new ParameterVector(model.Parameters().ToArray());

IList<Learner> parameterLearners = new List<Learner>()
{
    CNTKLib.AdaDeltaLearner(parameterVector, 
        schedule, 
        rho: 0.95, 
        epsilon: 1e-8,
        additionalOptions: options)
};
```

Python version has faster convergence rates, not only on this particular network. Is there a way I can achieve the same performance on C#?

**Comment by KeDengMS** [comment]
@tangyuq Seems like a bug in Python/C# inconsistency in [TrainingParameterSchedule](https://github.com/Microsoft/CNTK/blob/master/Source/CNTKv2LibraryDll/API/CNTKLibrary.h#L4925). By default, Python uses per-sample learning rate (minibatchSize == 1), while C# takes the default value of 0.

As a workaround, please manually specify the minibatchSize in TrainingParameterSchedule to 1 in your C# code to get the same convergence as Python.

**Comment by DomnenkoB** [comment]
@KeDengMS setting minibatchSize to 1 improved the convergence rate, but still, Python is faster

**Comment by KeDengMS** [comment]
Weight initialization may also affect convergence. Please try use the same init value. You may also dump computation outputs of each node/parameter using [set_computation_network_trace_level](https://cntk.ai/pythondocs/cntk.debugging.debug.html#cntk.debugging.debug.set_computation_network_trace_level) for a side-by-side comparison.

**Comment by tangyuq** [comment]
The C# APIs are still in their preliminary status not as matured as the Python APIs. In our Python APIs, on top of the core engine, we have done a lot more on the default hyper-parameters for the learners. For adadelta, there are Gaussian_noise_injection, l1 and l2 regularization, and so on. The C# training APIs are still lacking behind for this. 

In the meantime, you can copy the hyper-parameters used in Python APIs at:  https://github.com/Microsoft/CNTK/blob/master/bindings/python/cntk/learners/__init__.py#L803

**Comment by kkc0923** [comment]
@KeDengMS How about other learners? I used Adam with lr and momentum in C#, what default value is used for variance_momentum? In python, it seems that momentum_schedule_per_sample(0.9999986111120757), but I can't find any information of default value for C# learners. Should I manually set all of parameters to same value to the source in the link of @tangyuq to gain same training performance?

**Comment by DomnenkoB** [comment]
@tangyuq I have already specified some of learning parameters in the options

`
var options = new AdditionalLearningOptions()
{
    l1RegularizationWeight = 0,
    l2RegularizationWeight = 0,
    gradientClippingThresholdPerSample = double.PositiveInfinity,
    gradientClippingWithTruncation = true,
    // dictOptions = ???
};
`

However, there is `dictOptions` parameter. Is there a way I can specify Gaussian noise injection for instance?

**Comment by tangyuq** [comment]
I am not an expert on the C# API size.  I guess that new AdditionalLearningOptions() { ... gaussianNoiseInjectionStdDev = ...} should be the way to specify it. 

@liqunfu Do you have any hints?
